### PR TITLE
Adding support for PLIC priority threshold register

### PR DESCRIPTION
**Pull request by rohanjainRS**
### **Summary**
- Adding support to be able to read/write to priority threshold register such that a certain interrupt will only be executed if a source's priority >= a context's priority threshold.
- Changes made to src/Infrastructure/src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs include function AddContextPriorityThresholdRegisterPLIC() to create priority threshold registers and support read/write abilities. 
- Changes made to src/Infrastructure/src/Emulator/Cores/RiscV/PLIC/IrqContext.cs include creating variable for priority threshold, getter/setter function, and functionality for comparing source's priority to context's priority threshold in AcknowledgePendingInterrupt(). In default constructor, set this.priorityThreshold to 0 (ie capture interrupts of all priorities).
- Changes made to src/Infrastructure/src/Emulator/Cores/RiscV/PlatformLevelInterruptController.cs include creation of priority threshold registers for each context by calling AddContextPriorityThresholdRegisterPLIC() for each context. Also changed addresses of some registers to match those in Gemini code.

### **Testing**
- Tested with a sample program using UART interrupt from ATCUART100 peripheral: 
- The priority threshold of the context and priority of the UART interrupt were testing by changing variables in gemini_interrupt.c. 
- Set PLIC_REGS->properties_context[0].priority_threshold in subsystem_enable_interrupts() to set priority register for context 0. 
- Set PLIC_REGS->priority[irq_source - 18] in irq_map_unmask_and_enable_in_controller() to set priority of UART source.

-  UART will print an ISR message if interrupt is taken, along with corresponding log messages:

image

- If UART interrupt is not taken, test program will continue as usual and corresponding log messages will give details:

image

### CODE REVIEW

File: src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs [hunk 1]
```diff
@@ -66,6 +67,7 @@ public void Reset()
         public void WriteDoubleWord(long offset, uint value)
         {
             registers.Write(offset, value);
+            //Console.WriteLine("Debugging, PLICbase.cs: WriteDoubleWord(): offset = 0x{0:x8}, value = 0x{1:x8}", offset, value); //debugging
```
Comment (rahul-r-shah) [inline src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs]: Remove the commented log

File: src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs [hunk 2]
```diff
@@ -108,6 +110,7 @@ protected void RefreshInterrupts()
 
         protected void AddContextClaimCompleteRegister(Dictionary<long, DoubleWordRegister> registersMap, long offset, uint contextId)
         {
+            //Console.WriteLine("Debugging, AddContextClaimCompleteRegister(): offset = 0x{0:x8}, contextID = {1}", offset, contextId); //debugging
```
Comment (rahul-r-shah) [inline src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs]: Remove the commented log

File: src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs [hunk 3]
```diff
@@ -132,7 +138,7 @@ protected void AddContextClaimCompleteRegister(Dictionary<long, DoubleWordRegist
         protected void AddContextEnablesRegister(Dictionary<long, DoubleWordRegister> registersMap, long address, uint contextId, int numberOfSources)
         {
             var maximumSourceDoubleWords = (int)Math.Ceiling((numberOfSources + 1) / 32.0) * 4;
-
+            //Console.WriteLine("Debugging, AddContextEnablesRegister(): maximumSourceDoubleWords = 0x{0:x8}, address = 0x{1:x8}, contextID = {2}", maximumSourceDoubleWords, address, contextId); //debugging
```
Comment (rahul-r-shah) [inline src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs]: Remove the commented log

File: src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs [hunk 4]
```diff
@@ -165,6 +171,28 @@ protected void AddContextEnablesRegister(Dictionary<long, DoubleWordRegister> re
             }
         }
 
+        //adapted from OpenTitan_PlatformLevelInterruptController.cs
+        protected void AddContextPriorityThresholdRegisterPLIC(Dictionary<long, DoubleWordRegister> registersMap, long offset, uint contextId)
+        {
+            //Console.WriteLine("Debugging, AddContextPriorityThresholdRegister(): offset = 0x{0:x8}, contextID = {1}", offset, contextId); //debugging
```
Comment (rahul-r-shah) [inline src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs]: Remove the commented log

**Review by rahul-r-shah — CHANGES_REQUESTED** [review]
Some comments to be removed

File: src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs [hunk 5]
```diff
@@ -44,6 +44,7 @@ public PlatformLevelInterruptControllerBase(int numberOfSources, int numberOfCon
 
         public uint ReadDoubleWord(long offset)
         {
+            //Console.WriteLine("Debugging, PLICbase.cs: ReadDoubleWord(): offset = 0x{0:x8}", offset); //debugging
```
Comment (jorge-rs) [inline src/Emulator/Cores/RiscV/PLIC/PlatformLevelInterruptControllerBase.cs]: Delete this one too

**Review by jorge-rs — CHANGES_REQUESTED** [review]
From my side, changes look good to enable a better PLIC behavior in the model. Please, remove remaining debug messages which are as currently comments

**Review by jorge-rs — APPROVED** [review]

**Review by jorge-rs — APPROVED** [review]
LGTM

**Pull request merged**